### PR TITLE
fix(core): select by typing on a focused Selector (#3764)

### LINKED ISSUE #3764: [Bug] Selector component doesn't allow keyboard to select by typing
### Description

If I focus a Selector, I can't just press a key to select one of the items. For example if I have a state picker, I can't tab into it and press "C" to select "CA". I have to actually interact with the menu, which is annoying. This makes it hard to replace native selectors fully.

### Reproduction

1. Render <Selector> with some options
2. Tab into it so it's focused
3. Press a character that matches one of the options
4. I would expect that option to be selected

### Astryx Version

@astryxdesign/core@0.14

### Environment

OS X

### PR DESCRIPTION
**Pull request by AKnassa**
Fixes #3764

## What this does

You can now pick an option in a Selector just by typing. Focus the dropdown with Tab, press "C", and it selects "CA" — no need to open the menu and hunt for the option first. This is how the browser's own dropdowns have always worked, and Selector now matches that, so it can genuinely replace them.

## Why

The person who reported this was building a state picker. They could tab to it, but then had to stop and interact with the menu to choose anything — which makes a form slow to fill in by keyboard and made Selector a poor swap for a native dropdown.

## What changed

- **Type to pick.** Typing a letter on a closed Selector selects the first matching option. The menu stays shut, just like a native dropdown.
- **Press again to cycle.** If several options start with the same letter, pressing it again moves to the next one — Chicago, then Cleveland, then Columbus, then back around.
- **Spaces work.** Typing "new y" gets you to "New York" instead of the space opening the menu. The typed text is forgotten after a short pause, so the next thing you type starts fresh.
- **Screen readers hear it.** Because the value changes without the menu opening or focus moving, nothing would otherwise announce the new choice — so it's now announced politely.
- **Searchable Selectors too.** With a search box enabled, typing on the closed dropdown opens it and drops what you typed straight into the search field, without losing any characters.
- **Selector-only by design.** The matching lives in Selector's own use of the shared `useTypeahead` hook, not in the shared combobox. CommandPalette and MultiSelector get nothing extra, and the palette is already a search input, so it needs no opt-out.

Menu-open behavior is unchanged for the user: typing walks the highlight and Enter confirms. The same shared hook now drives it, moving the highlight via `aria-activedescendant`.

## How to see it

In Storybook, open **Core → Selector → Default**, click once outside the dropdown, then press Tab to focus it and type a letter. The value changes with the menu closed.

## Notes for reviewers

- Typeahead is a Selector-level composition of the shared `useTypeahead` hook, run ahead of the combobox key handler. Nothing was added to `useCombobox`, so CommandPalette and MultiSelector are untouched.
- Announcements use `useAnnounce` rather than local live-region state.
- Adopting the shared hook surfaced and fixed three bugs in it: a single-character search now starts *after* the current item (native `<select>` / APG `getIndexByLetter`); a `-1` start no longer wraps onto the last item; and a ctrl/meta-chorded Space no longer appends a raw space that poisons the buffer. DropdownMenu, ContextMenu, and NavHeadingMenu share the hook and benefit.
- The typeahead anchor and the duplicate-select guard read the optimistic value rather than the raw `value` prop, so a pending `changeAction` can't strand cycling on the first match or re-fire the action for a value already committed.
- The typed-text buffer resets on select, close, and clear, so a stale prefix can't poison the next keystroke.
- Unit tests cover the closed and open paths, cycling, spaces, disabled-skip, sections, and the shared-hook edge cases (chord-Space, null label slots, stale/negative index).
- Follow-up, not in this PR: `useAnnounce` mounts its live region on `document.body`, so a Selector inside a modal Dialog won't announce — a limitation shared by every current `useAnnounce` consumer, not specific to this change.

### CODE REVIEW

File: packages/core/src/CommandPalette/CommandPalette.tsx [hunk 1]
```diff
@@ -328,6 +328,11 @@ export function CommandPalette<T extends SearchableItem = SearchableItem>({
     selectableItems,
     value,
     isOpen: true, // Always "open" from combobox POV — the dialog handles visibility
+    // The palette's input already filters the items, and it forwards every
+    // keystroke here. Prefix typeahead on top of that would chase the same
+    // text and drag the highlight — and what Enter commits — off the user's
+    // pick, so the input stays the only thing typing drives.
+    hasTypeahead: false,
```
Comment (cixzhang) [inline packages/core/src/CommandPalette/CommandPalette.tsx]: This option probably doesn't make sense for the command palette since it's already a search input form of typeahead so it naturally performs this behavior. We'd only really need the behavior on selectors.

File: packages/core/src/Selector/Selector.tsx [hunk 2]
```diff
@@ -718,6 +717,17 @@ export function Selector<T extends SelectorOptionType>(
     }
   }, [onChange, changeAction, startTransition, setOptimisticValue]);
 
+  // Announces a value committed by typing on the closed trigger. That path
+  // changes the selection without opening the popup or moving focus, so no
+  // other event prompts assistive tech to re-read the trigger.
+  const [liveMessage, setLiveMessage] = useState('');
```
Comment (cixzhang) [inline packages/core/src/Selector/Selector.tsx]: Hmm it can be finicky to manage this manually but using state seems odd. Are you able to use useAnnounce instead?

File: packages/core/src/Selector/Selector.tsx [hunk 3]
```diff
@@ -729,32 +739,57 @@ export function Selector<T extends SelectorOptionType>(
...
+            input.setSelectionRange(input.value.length, input.value.length);
+          }
         });
       }
     }, [popover, hasSearch]),
     onClose: popover.hide,
     onSelect: useCallback(
       (newValue: string) => {
         onChange?.(newValue);
+        if (!popover.isOpen) {
```
Comment (cixzhang) [inline packages/core/src/Selector/Selector.tsx]: Hmm it should be possible to have the typeahead behavior when it's open (it should move focus to the item that matches) so we shouldn't need this gate.

File: packages/core/src/Selector/hooks.ts [hunk 4]
```diff
@@ -171,14 +195,32 @@ export function useCombobox({
...
+  const typeaheadRef = useRef('');
   const typeaheadTimeoutRef = useRef<ReturnType<typeof setTimeout> | undefined>(
     undefined,
   );
 
+  // Drop the buffer at every interaction boundary — committing, closing, and
+  // clearing all end the "word" being typed. Without this, a stale prefix
+  // survives for up to TYPEAHEAD_RESET_MS and poisons the next keystroke:
+  // selecting Dog then typing "c" would search for "dc" and match nothing.
+  const resetTypeahead = useCallback(() => {
```
Comment (cixzhang) [inline packages/core/src/Selector/hooks.ts]: Are we able to reuse useTypeahead instead of rolling something separately?

**Review by cixzhang — CHANGES_REQUESTED** [review]
Thank you for working on this! I left some comments mainly to reuse existing a11y hooks so we're not applying behaviors inconsistently.

**Comment by HelloOjasMutreja** [comment]
Dug into both open review questions against the current diff — concrete answers on both:

## `useTypeahead` reuse (hooks.ts)

`packages/core/src/hooks/useTypeahead.ts` already implements almost exactly what `findTypeaheadMatch`/`handleTypeahead` roll here: buffered printable-char matching, reset-after-inactivity via a ref+timeout, cycling to the *next* match on a repeated single character (APG `getIndexByLetter` behavior), and a caller-supplied `isDisabled` skip — the same four things this PR's custom version does. The composable shape (`onMatch(index)`, caller decides what to do with it) maps directly onto the open/closed branch this PR needs:

```ts
const typeahead = useTypeahead({
  getItemLabels: () => selectableItems.map(i => i.label ?? i.value),
  onMatch: (index) => {
    if (isOpen) {
      setHighlightedIndex(index);
    } else {
      const item = selectableItems[index];
      if (item.value !== value) onSelect?.(item.value);
    }
  },
  getCurrentIndex: () => (isOpen ? highlightedIndex : findSelectedIndex()),
  resetMs: 500, // matches TYPEAHEAD_RESET_MS here today (hook defaults to 750)
  isDisabled: (index) => selectableItems[index]?.disabled,
});
```

`useTypeahead.onKeyDown` also already handles the "space mid-buffer extends the query" case internally (`isSpaceMidType`), so the separate space-handling branch in the outer `onKeyDown` switch could likely delegate to it too instead of calling `handleTypeahead(' ')` directly. `reset()` maps onto the existing `resetTypeahead()` call sites (Escape, clear, close). Net: this should drop `findTypeaheadMatch` and most of `handleTypeahead`'s bookkeeping (~50 lines) with no behavior change — the existing typeahead test suite here should keep passing unchanged since it's an internal swap, not a behavior change.

## `useAnnounce` reuse (Selector.tsx)

`packages/core/src/hooks/useAnnounce.ts` is the existing pattern for exactly this — an imperative `announce(message)` backed by a **singleton, persistently-mounted** live region. Its own doc comment calls out why that matters: "many screen readers will not announce content injected into a live region that is created together with its content ('born with content')." The `<VisuallyHidden role="status" aria-live="polite">{liveMessage}</VisuallyHidden>` mounted here is created alongside its content on first render, which is the exact case that hook exists to avoid. Swapping to `useAnnounce()` + calling `announce(committed.label ?? newValue)` in the `onSelect` closed-trigger branch should be a drop-in replacement for the `liveMessage` state + the inline live region markup.

Happy to leave this as reference if useful — not trying to take over the branch, just wanted to give the open questions a concrete answer since they'd been sitting a while.

File: packages/core/src/CommandPalette/CommandPalette.tsx (hunk 1 above)
Comment (AKnassa) [reply packages/core/src/CommandPalette/CommandPalette.tsx]: `hasTypeahead` is gone. Typeahead lives in the shared `useTypeahead` hook now, so CommandPalette gets it as a search input with no opt-out to carry.

File: packages/core/src/Selector/Selector.tsx (hunk 2 above)
Comment (AKnassa) [reply packages/core/src/Selector/Selector.tsx]: Switched to `useAnnounce`. The `liveMessage` state and the inline `VisuallyHidden` region are removed.

File: packages/core/src/Selector/Selector.tsx (hunk 3 above)
Comment (AKnassa) [reply packages/core/src/Selector/Selector.tsx]: Gate removed. Open, it walks the highlight via `aria-activedescendant`; closed, it commits like a native select. It moves the virtual cursor rather than DOM focus, since the trigger keeps focus and the popup is the `role="listbox"`, so that is the combobox pattern here.

File: packages/core/src/Selector/hooks.ts (hunk 4 above)
Comment (AKnassa) [reply packages/core/src/Selector/hooks.ts]: Done. The separate matcher is deleted and it composes the shared `useTypeahead` ahead of the combobox key handler; `reset` fires on close and on clear so no stale prefix survives. Adopting it surfaced a few off-by-one bugs in the shared hook (single-char offset and the `-1` start), fixed in this PR, so DropdownMenu/ContextMenu/NavHeadingMenu benefit too. More detail in the PR description.

**Comment by AKnassa** [comment]
Thanks @HelloOjasMutreja. This is the direction the PR landed on, reusing the shared `useTypeahead` hook. Kept the 750ms reset (the shared default) rather than 500 so typeahead timing stays consistent across menus and selectors.

**Review by cixzhang — APPROVED** [review]
Thanks this is looking good now

**Comment by cixzhang** [comment]
Hi @AKnassa — this had drifted ~306 commits behind `main` and was conflicting, so I merged `main` into your branch on your behalf to clear it (a merge commit, no rebase and no force-push, so your history is untouched).

Both conflicts were in test files only; production code merged clean:

- `packages/core/src/Selector/Selector.test.tsx` — `main` added a `generateThemeTestCSS` helper next to a `politeRegion()` helper, while your branch had moved `politeRegion` further down the file. Kept `main`'s `generateThemeTestCSS` and your relocated `politeRegion`, dropping only the duplicate definition.
- `packages/core/src/CommandPalette/CommandPalette.test.tsx` — your "does not move the highlight while typing in the search input" test landed at the same spot as `main`'s new `screen reader announcements` describe block. Kept both.

Verified on the merged tree: no test cases lost (Selector 108, CommandPalette 20 — the exact union of both sides), `tsc --noEmit` clean, ESLint clean on the touched files, and 400/400 passing across Selector, useTypeahead, DropdownMenu, CommandPalette, ContextMenu, MultiSelector and NavHeading.

Nothing needed from you — just flagging that the branch moved.

**Comment by AKnassa** [comment]
Thank you so much!
